rust_build_utils/msvc.py

```diff
@@ -39,60 +42,74 @@ def activate_msvc(
     direct_pass_arch (optional): if True, the `arch` value will be passed to vcvarsall.bat
       without appending the 'amd64_' prefix.
 
+    Returns:
+    envrinmental variables and their original values that were modified by vcvarsall.bat script
+
     Example usage:
-    with activate_msvc('arm64'):
-        print(subprocess.run("link"))
-        print(subprocess.run("cl"))
+    orig_env = activate_msvc('arm64')
+    print(subprocess.run("link"))
+    print(subprocess.run("cl"))
+    deactivate_msvc(orig_env)
     """
 
     # Sample location of vcvarsall script:
     # "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\Build\vcvarsall.bat"
     # We begin by finding microsoft visual studio installation.
-    mvs_path = Path(r"C:\Program Files\Microsoft Visual Studio")
-    if not mvs_path.is_dir():
+    mvs_paths = [
+        Path(r"C:\Program Files (x86)\Microsoft Visual Studio"),
+        Path(r"C:\Program Files\Microsoft Visual Studio"),
+    ]
+    if not any(p.is_dir() for p in mvs_paths):
         raise Exception(
             "Microsoft Visual Studio might not be installed. Was looking in '{}'".format(
-                mvs_path
+                mvs_paths
             )
         )
 
+    edition_preferences = [
+        "BuildTools",
+        "Enterprise",
+        "Professional",
+        "Community",
+        "Preview",
+    ]
     # Multiple versions and multiple editions might be installed, so we iterate over versions installed.
     # If version preference is given, we filter only matching versions, othervise the highest version is picked.
     mvs_versions = sorted(
         [
             v
-            for v in mvs_path.iterdir()
-            if v.is_dir()
-            and (version_preference is None or v.name == version_preference)
+            for v in itertools.chain.from_iterable(p.iterdir() for p in mvs_paths)

```

Comment:
This is one gnarly way to write a double nested loop :D At this point I'm having trouble to comprehend whats happening here.

Reply:
Maybe the checking part could be separated into a local function? But this way the ifs are inverted as guard clauses, so it makes it hard to read as well :/
```
def check_version(v):
  if not v.is_dir():
    return None
 ..
```